src/optimistic_transaction_db.rs

```diff
@@ -0,0 +1,259 @@
+
+            unsafe {
+                db = ffi_try!(ffi::rocksdb_optimistictransactiondb_open_column_families(
+                    opts.inner,
+                    cpath.as_ptr(),
+                    cfs.len() as c_int,
+                    cf_names.as_mut_ptr(),
+                    cf_opts.as_mut_ptr(),
+                    cf_handles.as_mut_ptr(),
+                ));
+            }
+
+            for handle in &cf_handles {
+                if handle.is_null() {
+                    return Err(Error::new(
+                        "Received null column family \
+                         handle from DB."
+                            .to_owned(),
+                    ));
+                }
+            }
+
+            for (n, h) in cfs.iter().zip(cf_handles) {
+                cf_map
+                    .write()
+                    .map_err(|e| Error::new(e.to_string()))?
+                    .insert(n.name.clone(), h);
+            }
+        }
+
+        if db.is_null() {
+            return Err(Error::new("Could not initialize database.".to_owned()));
+        };
+
+        let path_buf = path.to_path_buf();
+
+        let base_db = unsafe {
+            let inner = ffi::rocksdb_optimistictransactiondb_get_base_db(db);
+            DB {
+                inner,
+                cfs: cf_map,
+                path: path_buf.clone(),
+                is_base_db: true,
+            }
+        };
+
+        Ok(OptimisticTransactionDB {
+            inner: db,
+            path: path_buf,
+            base_db,
+        })
+    }
+
+    /// Begins a new optimistic transaction.
+    pub fn transaction(
+        &self,
+        write_options: &WriteOptions,
+        optimistic_tx_options: &OptimisticTransactionOptions,
+    ) -> Transaction {
+        unsafe {
+            let inner = ffi::rocksdb_optimistictransaction_begin(
+                self.inner,
+                write_options.inner,
+                optimistic_tx_options.inner,
+                ptr::null_mut(),
+            );
+            let snapshot = if optimistic_tx_options.set_snapshot {
+                Some(ffi::rocksdb_transaction_get_snapshot(inner))
+            } else {
+                None
+            };
+            Transaction::new(inner, snapshot)
+        }
+    }
+
+    /// Begins a new optimistic transaction with default options.
+    pub fn transaction_default(&self) -> Transaction {
+        let write_options = WriteOptions::default();
+        let optimistic_transaction_options = OptimisticTransactionOptions::new();
+        self.transaction(&write_options, &optimistic_transaction_options)
+    }
+
+    // Get database path
+    pub fn path(&self) -> &Path {
+        &self.path.as_path()
+    }
+
+    /// Get base DB
+    pub fn get_base_db(&self) -> &DB {
+        &self.base_db
+    }
+}
+
+impl Drop for OptimisticTransactionDB {
+    fn drop(&mut self) {
+        self.base_db.drop_base_db();
```

Comment:
Is this necessary? Won't the `base_db` field get `Drop`ped because `DB` implements `Drop`?

Reply:
base_db automatically Drop later than OptimisticTransactionDB

Reply:
Ah, the ordering of drop matters.  I really don't like having an `is_base_db` field or the altered `drop` function for `DB`.  If you change `base_db` to an `Option<DB>`, you can use `Option<T>::take` in `OptimisticTransactionDB`'s `Drop` impl to change the ordering of the drop. As an example, the code snippet below produces the following output:

```
Dropping inner: Hello from Drop!
Dropping outer!
```

```
fn main() {
  let _ = Outer {
    inner: Some(Inner { message: "Hello from Drop!".to_string() })
  };
}

struct Outer {
  inner: Option<Inner>
}

struct Inner {
  message: String
}

impl Drop for Outer {
  fn drop(&mut self) {
    drop(self.inner.take());

    println!("Dropping outer!");
  }
}

impl Drop for Inner {
  fn drop(&mut self) {
    println!("Dropping inner: {}", self.message);
  }
}
```

Reply:
but how to replace `is_base_db` .
rocksdb_close and rocksdb_optimistictransactiondb_close_base_db are different.